### ISSUE TITLE
Feature request for fixing overdispersion parameter in NegativeBinomial distribution

### ISSUE THREAD

**Issue by minhuangr** [ManuelStapper/CountTimeSeries.jl#8]
Thanks for the implementation of the nice package. 

I hope you can add one more feature for the NegativeBinomial distribution. I would like to have the option to fix the overdispersion parameter ϕ instead of estimating it. For example, I would like to have the conditional Geometric distribution (ϕ = 1) or some other fixed ϕ NegativeBinomial distribution.   

Another side note here: I occasionally find that INARCH does not converge while INGARCH always does. Do you have any insight on this? Thanks.

**Comment by ManuelStapper** [comment]
Hi there, thanks for the feedback! Estimation with fixed parameter(s) seems like a great idea, maybe not only for the overdispersion parameter. I'm gonna take a look where/how it is implemented best and get back to you. 

Regarding the convergence issue: I would have assumed that the default maximum number of evaluations is sufficient for the optimization algorithms to converge. Do you have any additional background when it does not? Do you see it in real world data applications or simulated data? What are the settings you use? Any additional info would help, replication code would be awesome =)

**Comment by minhuangr** [comment]
Thank you for the quick reply.

The convergence problem occured in some loops to find the best model selected by HQIC. INGARCH never failed. I traced one possible data set but unfortunately it seems that you need to fit it third times to see the error message (during the loop, I did not refit the model, just error bumped) :

```
using CountTimeSeries
x = [14  6  19  7  6  17  7  3  6  15  14  8  2  4  5  1  1  17  16  5  5  8  20  3  2  2  10]
xdata = reshape(x,size(x,2))
model = Model(distr = "NegativeBinomial", pastObs =[1,3], pastMean = [])
res = fit(xdata, model, MLESettings(xdata, model, maxEval = 1e+9))
res = fit(xdata, model, MLESettings(xdata, model, maxEval = 1e+9))
res = fit(xdata, model, MLESettings(xdata, model, maxEval = 1e+9))
```

Error messages:

ERROR: DomainError with NaN:
NegativeBinomial: the condition zero(p) < p <= one(p) is not satisfied.

INARCH(6) also failed if I tried to fit it three times.

Some othe error messages during the loops:

ERROR: AssertionError: isfinite(phi_d) && isfinite(gphi)

**Comment by ManuelStapper** [comment]
Thanks for the replication code!

Good news: I found the bug that caused the error. In an internal function specifically for INARCH models, there was a problem when using non-consecutive orders as in the example you provided. I fixed it and it should work now when you re-install the package directly from GitHub.

When I'm done including the fixed parameter estimation, I'll bump the package to a new version. I will let you know in this issue, so let's keep it opened.

Thanks again!

**Comment by minhuangr** [comment]
Thanks for the quick fix. It works.

**Comment by ManuelStapper** [comment]
Hello again,
I experimented a little with restricted estimation and in the end decided to include the option to restrict any parameter(s). This can now be done by the keyword argument "restr" in the fit() function. Here's a MWE to fix the overdispersion parameter to one:

model = Model(pastObs=1, distr="NegativeBinomial")
y = simulate(1000, model, [10, 0.5, 1])[1]
setting = MLESettings(y, model, ci = true)
fit(y, model, setting, restr = ["ϕ" => 1])

If you want to include further restrictions, the argument "restr" accepts vectors of pairs, so you could do something like
fit(y, model, setting, restr = ["β0" => 10, "ϕ" => 1])

If you encounter any bugs during your simulations, please let me know.

Thanks again, for the feedback!

**Comment by minhuangr** [comment]
Thanks for the requested new feature. I did not see any problem yet. But your codes recalled one problem I have found before. Your CI is highly unstable which can barely be used.

```
modelt = Model(pastObs=3, distr="NegativeBinomial")
y = simulate(1000, modelt, [10, 0.5, 1])[1]
model = Model(pastObs=1:3, distr="NegativeBinomial")
setting = MLESettings(y, model)#, ci = true
fit(y, model, setting)
setting = MLESettings(y, model, ci = true)
fit(y, model, setting)
```

ERROR: ArgumentError: matrix contains Infs or NaNs

Another side request, for the restricted model fitting, can printResults show that it is fitting a restricted model?

**Comment by ManuelStapper** [comment]
Thanks again for the feedback.

In the code you provided, I assume the Hessian can't be computed numerically, because one of the estimates is on the boundary of parameter space. Might be a good idea to include for example Bootstrapping as an alternative method.

I will fix the console output to display the restricted parameters. 

All the best.

**Comment by minhuangr** [comment]
Yep, I understand some parameters may be on the boundary. I do appreciate that you have made many checks during your model setups, fits, etc. I would expect to have some message or warnings in this situation instead of force quit of the function. Anyway, I will use try...catch to handle it myself.

Thanks again for the quick response and display request.
